src/libostree/ostree-repo.c

```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+
+  if (lock != NULL)
+    {
+      guint stack_len;
+      const char *cur_state_name;
+
+      if (lock->stack)
+        {
+          int cur_state;
+
+          stack_len = g_queue_get_length (lock->stack);
+          if (stack_len == 0)
+            {
+              cur_state = LOCK_UN;
+              cur_state_name = "unlocked";
+            }
+          else
+            {
+              cur_state = GPOINTER_TO_INT (g_queue_peek_head (lock->stack));
+              cur_state_name = (cur_state == LOCK_EX) ? "exclusive" : "shared";
+            }
+        }
+      else
+        {
+          stack_len = 0;
+          cur_state_name = "null";
+        }
+
+      g_debug ("Free lock: state=%s, depth=%u", cur_state_name, stack_len);
+      g_queue_free (lock->stack);
+      if (lock->fd >= 0)
+        {
+          g_debug ("Closing repo lock file");
+          (void) close (lock->fd);
+        }
+      g_free (lock);
+    }
+}
+
+/* Wrapper to handle flock vs OFD locking based on GLnxLockFile */
+static gboolean
+do_repo_lock(int fd,
+             int flags)
+{
+  int res;
+
+#ifdef F_OFD_SETLK
+  struct flock fl = {
+    .l_type = (flags & ~LOCK_NB) == LOCK_EX ? F_WRLCK : F_RDLCK,
+    .l_whence = SEEK_SET,
+  };
+
+  res = fcntl (fd, (flags & LOCK_NB) ? F_OFD_SETLK : F_OFD_SETLKW, &fl);
+#else
+  res = -1;
+  errno = EINVAL;
+#endif
+
+  /* Fallback to flock when OFD locks not available */
+  if (res < 0)
+    {
+      if (errno == EINVAL)
+        res = flock (fd, flags);
+      if (res < 0)
+        return FALSE;
+    }
+
+  return TRUE;
+}
+
+/* Wrapper to handle flock vs OFD unlocking based on GLnxLockFile */
+static gboolean
+do_repo_unlock(int fd,
+               int flags)
+{
+  int res;
+
+#ifdef F_OFD_SETLK
+  struct flock fl = {
+    .l_type = F_UNLCK,
+    .l_whence = SEEK_SET,
+  };
+
+  res = fcntl (fd, (flags & LOCK_NB) ? F_OFD_SETLK : F_OFD_SETLKW, &fl);
+#else
+  res = -1;
+  errno = EINVAL;
+#endif
+
+  /* Fallback to flock when OFD locks not available */
+  if (res < 0)
+    {
+      if (errno == EINVAL)
+        res = flock (fd, LOCK_UN | flags);
+      if (res < 0)
+        return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+push_repo_lock (OstreeRepo          *self,
+                OstreeRepoLockType   lock_type,
+                gboolean             blocking,
+                GError             **error)
+{
+  int flags = (lock_type == OSTREE_REPO_LOCK_EXCLUSIVE) ? LOCK_EX : LOCK_SH;
+  if (!blocking)
+    flags |= LOCK_NB;
+
+  GHashTable *lock_table = g_private_get (&repo_lock_table);
+  if (lock_table == NULL)
+    {
+      g_debug ("Creating repo lock table");
+      lock_table = g_hash_table_new_full (NULL, NULL, NULL,
+                                          (GDestroyNotify)free_repo_lock);
+      g_private_set (&repo_lock_table, lock_table);
+    }
+
+  OstreeRepoLock *lock = g_hash_table_lookup (lock_table, self);
+  if (lock == NULL)
+    {
+      lock = g_new (OstreeRepoLock, 1);
```

Comment:
Nitpick: I’d always use `g_new0()` for paranoia reasons.

Reply:
Heh. I actually did that intentionally for similar paranoid reasons. If `free_repo_lock` is called and `lock->fd == 0`, then you'd close `stdin`. So, I explicitly ordered the initialization below so you wouldn't do that. But I guess the compiler could do something different either way... I'll change it to `g_new0` so we don't accidentally free random memory.

Reply:
Fixed in #1343.